Public/Src/Cache/ContentStore/Library/Stores/FileSystemContentStoreInternal.cs

```diff
@@ -349,6 +354,195 @@ private Task Upgrade2To3(Context context)
+        {
+            try
+            {
+                // The cache entry is invalid if the size in content directory doesn't mach an actual size
+                if (ContentDirectory.TryGetFileInfo(expectedHash, out var contentFileInfo) && contentFileInfo.FileSize != fileInfo.Length)
+                {
+                    return (isValid: false, error: $"File size mismatch. Expected size is {contentFileInfo.FileSize} and size on disk is {fileInfo.Length}.");
+                }
+
+                // Or if the content doesn't match the hash.
+                var actualHashAndSize = await TryHashFileAsync(context, fileInfo.FullPath, expectedHash.HashType);
+                if (actualHashAndSize != null && actualHashAndSize.Value.Hash != expectedHash)
+                {
+                    // Don't need to add an expected hash into the error string because the client code will always put it into the final error message.
+                    return (isValid: false, error: $"Hash mismatch. Actual hash is {actualHashAndSize.Value.Hash}.");
+                }
+
+                return (isValid: true, error: string.Empty);
+            }
+            catch (IOException e)
+            {
+                _tracer.Warning(context, $"Content hash validation failed. Hash={expectedHash}, Error={e}.");
+
+                return (isValid: true, error: string.Empty);
+            }
+        }
+
+        private async Task<ContentHashWithSize?> TryHashFileAsync(Context context, AbsolutePath path, HashType hashType, Func<Stream, Stream> wrapStream = null)
+        {
+            // We only hash the file if a trusted hash is not supplied
+            using (var stream = await FileSystem.OpenAsync(path, FileAccess.Read, FileMode.Open, FileShare.Read | FileShare.Delete))
+            {
+                if (stream == null)
+                {
+                    return null;
+                }
+
+                using (var wrappedStream = (wrapStream == null) ? stream : wrapStream(stream))
+                {
+                    // Hash the file in  place
+                    return await HashContentAsync(context, wrappedStream, hashType, path);
+                }
+            }
+        }
+
+        private bool SelfCheckUpToDate()
+        {
+            if (!FileSystem.FileExists(_selfCheckFilePath))
+            {
+                return false;
+            }
+
+            // The format is: "SelfCheckEpoch|LastSelfCheckTime"
+            var contents = FileSystem.ReadAllText(_selfCheckFilePath);
+            var parts = contents.Split('|');
+            if (parts.Length != 2 || parts[0] != _settings.SelfCheckEpoch)
+            {
+                return false;
+            }
+
+            var reconcileTime = DateTimeUtilities.FromReadableTimestamp(parts[1]);
+            if (reconcileTime == null)
+            {
+                return false;
+            }
+
+            return reconcileTime.Value.IsRecent(Clock.UtcNow, _settings.SelfCheckFrequency);

```

Comment:
>reconcileTime.Value.IsRecent(Clock.UtcNow, _settings.SelfCheckFrequency); [](start = 19, length = 73)

Given that CASaaS is long lived, this will only make sense if SelfCheck is called inside a timer. Otherwise, the interval might be reached sometime after the call to SelfCheck.

Reply:
Agree with Lance. Maybe I am missing reason for this approach, though. Why not just create repeating Timer with specified interval? Within callback, simply exit if last callback is still running as determined by Interlocked.CompareExchange. Also turn off when shutdown/canceled. This seems simpler, AND it handles unusual case of empty cache without entering a tight loop.

---
In reply to: [282938941](https://github.com/microsoft/BuildXL/pull/289#discussion_r282938941) [](ancestors = 282938941)

Reply:
Let's discuss offline. Should be easier.

---
In reply to: [283074338](https://github.com/microsoft/BuildXL/pull/289#discussion_r283074338) [](ancestors = 283074338,282938941)